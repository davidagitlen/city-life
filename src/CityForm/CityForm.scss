.CityForm {
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  height: 100px;
  width: 100%;
  background-color: pink;

  & input {
    border-radius: 10px;
    border: none;
    height: 30px;
    width: 90%;
    margin: 10px;
    padding: 0px 0px 0px 10px;

    &[list]::-webkit-calendar-picker-indicator {
      display: none;
    }
  }

  & button {
      height: 30px;
      width: 90%;
      border-radius: 10px;
      margin: 10px;
  }

}