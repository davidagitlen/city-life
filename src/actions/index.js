export const setCityScores = (ordinal, array) => ({
  type: 'SET_CITY_SCORES',
  ordinal,
  array
});