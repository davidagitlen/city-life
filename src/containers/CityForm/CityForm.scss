.CityForm {
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  height: 150px;
  width: 100%;
  position: relative;
  background-color: ghostwhite;

  & input {
    border-radius: 10px;
    border: none;
    height: 30px;
    width: 90%;
    margin: 10px;
    padding: 0px 0px 0px 10px;

    &[list]::-webkit-calendar-picker-indicator {
      display: none;
    }
  }

  & label {
    position: relative;
    right: 33%;
    margin: 0px 0px 0px 10px;
  }

  & button {
      color: #2E2EC9;
      height: 30px;
      width: 90%;
      border-radius: 10px;
      margin: 10px 0px 10px 0px;

    &:hover {
      color: ghostwhite;
      background-color: #2E2EC9;
      cursor: pointer;
      transition: all 0.2s ease-in-out;
    }

    &:focus {
      color: ghostwhite;
      background-color: #2E2EC9;
      transition: all 0.2s ease-in-out;
    }

  }

  & p {
    margin: 10px 0px 10px 0px;
  }

}