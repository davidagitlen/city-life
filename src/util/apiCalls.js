export const fetchUrbanAreas = async () => {
  const url = 'https://api.teleport.org/api/urban_areas/';
  const response = await fetch(url);
  if (!response.ok) {
    throw new Error('There was a problem retrieving city data!');
  }
  const urbanAreasData = await response.json();
  return urbanAreasData;
}

export const fetchCityScores = async (partialUrl) => {
  const url = `${partialUrl}scores`;
  const response = await fetch(url);
  if (!response.ok) {
    throw new Error('There was a problem retrieving that city\'s scores!');
  }
  const cityData = await response.json();
  return cityData;
}

export const fetchCityImages = async (partialUrl) => {
  const url = `${partialUrl}images`;
  const response = await fetch(url);
  if (!response.ok) {
    throw new Error('There was a problem retrieving that city\'s images!');
  }
  const cityImages = await response.json();
  return cityImages;
}

export const findAdditionalData = async (city, originalName) => {
  const url = `https://api.teleport.org/api/cities/?search=${city}`;
  const response = await fetch(url);
  if (!response.ok) {
    throw new Error('There was a problem retrieving that city\'s additional data!');
  }
  const additionalCityData = await response.json();
  switch (originalName) {
    case 'Portland, ME' :
      return additionalCityData._embedded['city:search-results'][1]._links['city:item'].href;
    case 'Birmingham, AL' :
      return additionalCityData._embedded['city:search-results'][1]._links['city:item'].href;
    default :
      return additionalCityData._embedded['city:search-results'][0]._links['city:item'].href;
  }
}

export const fetchAdditionalData = async (href) => {
  const url = `${href}`
  const response = await fetch(url);
  if (!response.ok) {
    throw new Error('There was a problem parsing that city\'s additional data!')
  }
  const data = await response.json();
  return data;
}