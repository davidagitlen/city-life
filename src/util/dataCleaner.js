export const formatCityName = (name) => {
  switch (name) {
    case 'San Francisco Bay Area':
      return 'San Francisco';
    case 'Portland, ME':
      return 'Portland';
    case 'Portland, OR':
      return 'Portland';
    case 'Birmingham, AL':
      return 'Birmingham';
    case 'Minneapolis-Saint Paul':
      return 'Minneapolis';
    default:
      return name;
  }
};

export const formatAdditionalCityData = (data) => ({
  population: data.population.toLocaleString(),
  latitude: data.location.latlon.latitude,
  longitude: data.location.latlon.longitude,
  fullName: data.full_name,
  timeZone: data._links['city:timezone'].name,
  country: data._links['city:country'].name
});

export const formatChartData = (arrayOne, arrayTwo, indexes) => ({
  cityOneData: indexes.map(index => arrayOne[index]),
  cityTwoData: indexes.map(index => arrayTwo[index]),
});

export const isEmpty = (dataType, dataSet) => {
  switch(dataType) {
    case 'object':
      return dataSet.map(nestedValue => (!Object.keys(nestedValue).length)).includes(true);
    case 'array':
      return dataSet.map(nestedValue => (!nestedValue.length)).includes(true); 
    default:
      return true;
  }
}

