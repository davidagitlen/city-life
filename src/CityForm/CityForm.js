import React, { Component } from 'react';
import './CityForm.scss';
import { setCityScores, setCityDetails, setCityImages } from '../actions';
import { fetchUrbanAreas, fetchCityScores, findAdditionalData, fetchAdditionalData, fetchCityImages } from '../util/apiCalls';
import { formatCityName, formatAdditionalCityData } from '../util/dataCleaner';
import { connect } from 'react-redux';

class CityForm extends Component{
  constructor(){
    super();
    this.state = {
      city : '',
      cities : [],
      error : ''
    };
  }

  componentDidMount = async () => {
    try {
      const urbanAreas = await fetchUrbanAreas();
      const urbanAreasNames = urbanAreas._links['ua:item'];
      this.setState({cities: urbanAreasNames});
    } catch ({ message }) {
      this.setState({error : message}, () => {console.error(message)});
    }
  }

  handleCitySelection = e => {
    this.setState({[e.target.name]:e.target.value})
  }

  handleSubmitCity = async (e) => {
    e.preventDefault();
    this.getCityScores();
    this.getCityImages();
    this.handleAdditionalData();
  }

  getCityScores = async () => {
    const citySnippet = this.state.city.toLowerCase().replace(/,|\./g, '').replace(/\s/g, '-');
    try {
      const city = await fetchCityScores(`https://api.teleport.org/api/urban_areas/slug:${citySnippet}/`);
      const cityScores = city.categories;
      this.props.setCityScores(this.props.ordinal, cityScores);
    } catch ({ message }) {
      this.setState({ error: message }, () => { console.error(message) });
    }
  }

  getCityImages = async () => {
    const citySnippet = this.state.city.toLowerCase().replace(/,|\./g, '').replace(/\s/g, '-');
    try {
      const city = await fetchCityImages(`https://api.teleport.org/api/urban_areas/slug:${citySnippet}/`);
      const cityImages = {
        attribution: city.photos[0].attribution,
        images : city.photos[0].image
      };
      this.props.setCityImages(this.props.ordinal, cityImages)
    } catch ({ message }) {
      this.setState({ error: message}, () => { console.error(message)});
    } 
  }

  getAdditionalCityData = async () => {
    const originalName = this.state.city;
    // const cityName = this.formatCityName(this.state.city).replace(/,|\./, '')
    const cityName = formatCityName(originalName).replace(/,|\./, '');
    try {
      const city = await findAdditionalData(cityName, originalName);
      return city;
    } catch ({ message }) {
      this.setState({ error: message}, () => {console.error(message)});
    }
  }

  // formatCityName = (name) => {
  //   switch(name) {
  //     case 'San Francisco Bay Area' :
  //       return 'San Francisco';
  //     case 'Portland, ME' :
  //       return 'Portland';
  //     case 'Portland, OR' :
  //       return 'Portland';
  //     case 'Birmingham, AL' :
  //       return 'Birmingham';
  //     case 'Minneapolis-Saint Paul' :
  //       return 'Minneapolis';
  //     default :
  //       return name; 
  //   }
  // }

  handleAdditionalData = async () => {
    const href = await this.getAdditionalCityData();
    try {
      const additionalData = await fetchAdditionalData(href);
      // console.log('additionalData return', additionalData)
      // const formattedData = {
      //   population: additionalData.population.toLocaleString(),
      //   latitude: additionalData.location.latlon.latitude,
      //   longitude: additionalData.location.latlon.longitude,
      //   fullName: additionalData.full_name,
      //   timeZone: additionalData._links['city:timezone'].name,
      //   country: additionalData._links['city:country'].name
      // }
      const formattedData = formatAdditionalCityData(additionalData);
      // console.log(formattedData)
      this.props.setCityDetails(this.props.ordinal, formattedData);
    } catch ({ message }) {
      this.setState({ error : message});
    }
  }

  render() {
    const ordinal = this.props.ordinal;
    const cityOrdinal = ordinal === 'one' ? 'First' : 'Second';
    const cities = this.state.cities;
    const cityNames = cities.length ? cities.map(city => <option key={city.name} value={city.name} />) : null;
    return(
      <div className='CityForm'>
        <form>
          <p>{cityOrdinal} City</p>
          <label htmlFor='city'>City Name</label>
          <input 
            list='city-selection'
            name='city'
            id='city'
            type='text' 
            value={this.state.city}
            placeholder='Select a City'
            autoComplete='off'
            onChange={this.handleCitySelection}
            />
          <datalist id='city-selection'>
            {cityNames}
          </datalist>
          <button type='submit' disabled={!this.state.city} onClick={this.handleSubmitCity}>
            Select City
          </button>
        </form>
      </div>
    )
  }
}

export const mapStateToProps = state => ({
  cityInfo: state.cityInfo
});

export const mapDispatchToProps = dispatch => ({
  setCityScores: (ordinal, array) => dispatch(setCityScores(ordinal, array)),
  setCityDetails: (ordinal, details) => dispatch(setCityDetails(ordinal, details)),
  setCityImages: (ordinal, images) => dispatch(setCityImages(ordinal, images))
});

export default connect(mapStateToProps, mapDispatchToProps)(CityForm);