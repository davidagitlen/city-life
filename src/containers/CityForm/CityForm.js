import React, { Component } from 'react';
import './CityForm.scss';
import { bindActionCreators } from 'redux';
import {
  setCityScores, setCityDetails, setCityImages,
  requestCityScores, requestCityDetails, requestCityImages,
} from '../../actions';
import { fetchUrbanAreas, fetchCityScores, findAdditionalData, fetchAdditionalData, fetchCityImages } from '../../util/apiCalls';
import { formatCityName, formatAdditionalCityData } from '../../util/dataCleaner';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
const path = process.env.REACT_APP_URL;

export class CityForm extends Component{
  constructor(){
    super();
    this.state = {
      city : '',
      cities : [],
      error : ''
    };
  }

  componentDidMount = async () => {
    try {
      const urbanAreas = await fetchUrbanAreas();
      const urbanAreasNames = urbanAreas._links['ua:item'];
      this.setState({cities: urbanAreasNames});
    } catch ({ message }) {
      this.setState({error : message});
    }
  }

  handleCitySelection = e => {
    this.setState({[e.target.name]:e.target.value})
  }

  handleSubmitCity = async (e) => {
    e.preventDefault();
    this.getCityScores();
    this.getCityImages();
    this.handleAdditionalData();
  }

  getCityScores = async () => {
    const citySnippet = this.state.city.toLowerCase().replace(/,|\./g, '').replace(/\s/g, '-');
    this.props.requestCityScores(this.props.ordinal);
    try {
      const city = await fetchCityScores(`${path}/urban_areas/slug:${citySnippet}/`);
      const cityScores = city.categories;
      this.props.setCityScores(this.props.ordinal, cityScores);
    } catch ({ message }) {
      this.setState({ error: message });
    }
  }

  getCityImages = async () => {
    const citySnippet = this.state.city.toLowerCase().replace(/,|\./g, '').replace(/\s/g, '-');
    this.props.requestCityImages(this.props.ordinal);
    try {
      const city = await fetchCityImages(`${path}/urban_areas/slug:${citySnippet}/`);
      const cityImages = {
        attribution: city.photos[0].attribution,
        images : city.photos[0].image
      };
      this.props.setCityImages(this.props.ordinal, cityImages)
    } catch ({ message }) {
      this.setState({ error: message});
    } 
  }

  getAdditionalCityData = async () => {
    const originalName = this.state.city;
    const cityName = formatCityName(originalName).replace(/,|\./, '');
    try {
      const city = await findAdditionalData(cityName, originalName);
      return city;
    } catch ({ message }) {
      this.setState({ error: message});
    }
  }

  handleAdditionalData = async () => {
    const href = await this.getAdditionalCityData();
    this.props.requestCityDetails(this.props.ordinal);
    try {
      const additionalData = await fetchAdditionalData(href);
      const formattedData = formatAdditionalCityData(additionalData);
      this.props.setCityDetails(this.props.ordinal, formattedData);
    } catch ({ message }) {
      this.setState({ error : message});
    }
  }

  render() {
    const ordinal = this.props.ordinal;
    const cityOrdinal = ordinal === 0 ? 'First' : 'Second';
    const cities = this.state.cities;
    const cityNames = cities.length ? cities.map(city => <option key={city.name} value={city.name} />) : null;
    return(
      <div className='CityForm'>
        <form>
          <p>{cityOrdinal} City</p>
          <label htmlFor={`city-${ordinal}`}>City Name</label>
          <input 
            list='city-selection'
            name='city'
            id={`city-${ordinal}`}
            type='text' 
            value={this.state.city}
            placeholder='Select a City to Compare'
            autoComplete='off'
            onChange={this.handleCitySelection}
            />
          <datalist id='city-selection'>
            {cityNames}
          </datalist>
          <button type='submit' disabled={!this.state.city} onClick={this.handleSubmitCity}>
            Select City
          </button>
        </form>
      </div>
    )
  }
}

export const mapStateToProps = state => ({
  cityInfo: state.cityInfo
});

export const mapDispatchToProps = dispatch => bindActionCreators(
  {
    requestCityScores,
    setCityScores,
    requestCityDetails,
    setCityDetails,
    requestCityImages,
    setCityImages,
  },
  dispatch,
);

export default connect(mapStateToProps, mapDispatchToProps)(CityForm);

CityForm.propTypes = {
  cityInfo: PropTypes.object,
  ordinal: PropTypes.number,
  setCityDetails: PropTypes.func,
  setCityImages: PropTypes.func,
  setCityScores: PropTypes.func,
  requestCityDetails: PropTypes.func,
  requestCityImages: PropTypes.func,
  requestCityScores: PropTypes.func
}