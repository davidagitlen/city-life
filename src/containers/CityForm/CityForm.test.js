import React from 'react';
import { CityForm, mapStateToProps, mapDispatchToProps } from './CityForm';
import { setCityScores, setcityDetails, setCityImages } from '../../actions';
import { fetchUrbanAreas, fetchCityScores, findAdditionalData, fetchAdditionalData, fetchCityImages } from '../../util/apiCalls';
import { formatCityName, formatAdditionalCityData } from '../../util/dataCleaner'; 
import { shallow } from 'enzyme';

jest.mock('../../util/apiCalls');

describe('CityForm', () => {
  let wrapper, mockCityInfo;

  beforeEach(() => {
    mockCityInfo = {
      one: {
        details: {
          population: 1,
          latitude: 10,
          longitude: 10,
          fullName: 'Denver, Colorado!',
          timeZone: 'Denver!'
        },
        images: {
          attribution: {},
          images: {
            mobile: 'http://blurg.com',
            web: 'http://blurg.com/web'
          }
        },
        scores: [{ color: 'blue', name: 'Score', score_out_of_10: 10 }]
      },
      two: {
        details: {
          population: 2,
          latitude: 20,
          longitude: 20,
          fullName: 'Boulder, Colorado!',
          timeZone: 'Denver!'
        },
        images: {
          attribution: {},
          images: {
            mobile: 'http://blurg.com',
            web: 'http://blurg.com/web'
          }
        },
        scores: [{ color: 'red', name: 'Score', score_out_of_10: 10 }]
      }
    };

    fetchUrbanAreas.mockImplementation(() => {
      return Promise.resolve({
        urbanAreas: {
          _links: {
            'ua:item' : 'Denver'
          }
        }
      });
    });

    fetchCityScores.mockImplementation(() => {
      return Promise.resolve({
        categories: [{ color: 'blue', name: 'Score', score_out_of_10: 10 }]
      });
    });

    findAdditionalData.mockImplementation(() => {
      return Promise.resolve({});
    });

    fetchAdditionalData.mockImplementation(() => {
      return Promise.resolve({});
    });

    fetchCityImages.mockImplementation(() => {
      return Promise.resolve({});
    });

    wrapper = shallow(
      <CityForm 
        cityInfo={mockCityInfo}
        ordinal='one'
      />
    );
  });

  it('should match the snapshot with data passed in correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  describe('handleCitySelection', () => {

    it.only('should update city property of state with input value', () => {
      const mockEvent = {
        target: {
          name: 'city',
          value: 'Uncasville'
        }
      };

      const expected = {
        city: 'Uncasville',
        cities: [],
        error: ""
      };

      const wrapper = shallow(
        <CityForm 
          cityInfo={mockCityInfo}
          ordinal='one'
        />, {disableLifecycleMethods: true}
      )


      wrapper.instance().handleCitySelection(mockEvent);

      expect(wrapper.state()).toEqual(expected);
    });
  });

  describe('handleSubmitCity', () => {

    it('should invoke getCityScores, getCityImages, and handleAdditionalData', async () => {

      const mockEvent = {
        preventDefault: jest.fn()
      };
      
      wrapper.instance().getCityScores = jest.fn();
      wrapper.instance().getCityImages = jest.fn();
      wrapper.instance().handleAdditionalData = jest.fn();
      wrapper.instance().forceUpdate();

      await wrapper.instance().handleSubmitCity(mockEvent);

      expect(wrapper.instance().getCityScores).toHaveBeenCalled();
      expect(wrapper.instance().getCityImages).toHaveBeenCalled();
      expect(wrapper.instance().handleAdditionalData).toHaveBeenCalled();

    });
  });

  
});