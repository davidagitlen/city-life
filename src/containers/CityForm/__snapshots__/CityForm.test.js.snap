// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CityForm should match the snapshot with data passed in correctly 1`] = `
<div
  className="CityForm"
>
  <form>
    <p>
      First
       City
    </p>
    <label
      htmlFor="city-0"
    >
      City Name
    </label>
    <input
      autoComplete="off"
      id="city-0"
      list="city-selection"
      name="city"
      onChange={[Function]}
      placeholder="Enter a City Name"
      type="text"
      value=""
    />
    <datalist
      id="city-selection"
    />
    <button
      disabled={true}
      onClick={[Function]}
      type="submit"
    >
      Select City
    </button>
  </form>
</div>
`;
